### DISCUSSION TITLE
Question (Issue?) with Set-PnPListItemAsRecord

### DISCUSSION THREAD

**Discussion by renewtx** [pnp/powershell#2312] [Q&A]
Hi,

I am working on a PnP-based site provisioning script that uses a site configuration file (XML-based) extracted via  Get-PnPSiteTemplate and applies that template to a new site via Invoke-PnPSiteTemplate.  In the source site I have two libraries configured, along with a couple of views.  Features enabled on the source site include the Document ID and In Place Records Management features.

The application of the template to the target site works fine (no errors), and the libraries get created and the features are activated as expected.  As part of an extension to the provisioning I am now trying to programmatically update my predefined views to include the 'Declared Record' column ... however, this column is not available until AFTER the first time a record has been declared (it get's dynamically added to lists/libraries by the In  Place Rec Mgmt feature the first time you declare a record.

So what I was going to do it is (as part of my provisioning script) programmatically upload a file to my libraries, then use Set-PnPListItemAsRecord to declare that item as a record, then clear the record using Clear-PnPListItemAsRecord, and then delete the file.  Then I was going to update my views to include the additional column.  Unfortunately, when I call Set-PnPListItemAsRecord I get this error:

Set-PnPListItemAsRecord : Invalid field name. {3afcc5c7-c6ef-44f8-9479-3561d72f9e8e}  /sites/TargetSite17/KEEPit Document
At C:\temp\Novo\Provisioning.ps1:214 char:5
+     Set-PnPListItemAsRecord -List $ListName -Identity $ListItems.Id
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (:) [Set-PnPListItemAsRecord], ServerException
    + FullyQualifiedErrorId : EXCEPTION,PnP.PowerShell.Commands.RecordsManagement.SetListItemAsRecord

The GUID of the field it is referring to matches a GuID in the template XML that I use (which came from my template site). 

       <Field ID="{3AFCC5C7-C6EF-44f8-9479-3561D72F9E8E}" Name="_vti_ItemHoldRecordStatus" StaticName="_vti_ItemHoldRecordStatus" DisplayName="Hold and Record Status" SourceID="http://schemas.microsoft.com/sharepoint/v3" Group="Document and Record Management Columns" Type="Integer" Min="0" Max="32" Decimals="0" Indexed="FALSE" Hidden="TRUE" CanToggleHidden="FALSE" ShowInNewForm="FALSE" ShowInEditForm="FALSE" ShowInFileDlg="FALSE" ShowInDisplayForm="FALSE" Required="FALSE" Sealed="TRUE" ReadOnly="TRUE" />

In the log for the template application I also see that the field is (apparently) properly deployed

PowerShell_ISE.exe Information: 0 : 2022-09-02 17:02:17.1291	[Fields]	[16]	[Debug]	Updating field {3AFCC5C7-C6EF-44f8-9479-3561D72F9E8E} in site	3746ms	68df07d4-2a32-435f-a7b3-0614de36c6cc

However, Set-PnPListItemAsRecord apparently cannot find this field.  Any idea what is going on here?  Am I missing a step/component somewhere that is not obvious?

Note that I am able to use Set-PnPListItemAsRecord as described above to declare a record in the template site, so i know the template site is properly configured to allow declaring records ... just not the new sites generated via the template.

Any assistance/ideas appreciated.

Thanks

**Comment by FilipVGAgidens** [comment]
Hi,

I have the same issue. Were you able to fix it?

Kind Regards,
Filip,

**Reply by renewtx** [reply]
Hi -- sort of, by using a workaround.  If I recall, I had originally set up my site template libraries and then added a document, declared it as a record, then undeclared it and deleted it, before I captured the XML provisioning template.  But that seemed to cause issues.  So now i just create a library on the template site, grab the XML provisioning template, and then deploy to the target site.  Then, in my script, as part of provisioning, I programmatically add a document, declare/undeclare it, and then delete it to get the columns to get added.  That seems to work.

**Comment by FilipVGAgidens** [comment]
Hi, I solved it this way;

if($Error[0].Exception.Message -like "*3afcc5c7-c6ef-44f8-9479-3561d72f9e8e*"  ){
        $myfield = get-pnpField -connection $Myconnection -Identity "3afcc5c7-c6ef-44f8-9479-3561d72f9e8e" -InSiteHierarchy
        Add-PnPField -List $ListID -connection $Myconnection -Field $myfield
        write-host "Problem fixed"
    }
